staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go

```diff
@@ -266,6 +241,40 @@ func (s *Serializer) RecognizesData(data []byte) (ok, unknown bool, err error) {
 	return utilyaml.IsJSONBuffer(data), false, nil
 }
 
+func (s *Serializer) unmarshal(into runtime.Object, data, originalData []byte) error {
+	// If the deserializer is non-strict, return here.
+	if !s.options.Strict {
+		if err := kjson.UnmarshalCaseSensitivePreserveInts(data, into); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	var allStrictErrs []error
+	if s.options.Yaml {
+		// In strict mode pass the original data through the YAMLToJSONStrict converter.
+		// This is done to catch duplicate fields in YAML that would have been dropped in the original YAMLToJSON conversion.
+		// TODO: rework YAMLToJSONStrict to return warnings about duplicate fields without terminating so we don't have to do this twice.
+		_, err := yaml.YAMLToJSONStrict(originalData)
+		if err != nil {
+			allStrictErrs = append(allStrictErrs, err)
+		}
+	}
+
+	strictJSONErrs, err := kjson.UnmarshalStrict(data, into)
```

Comment:
Was just discussing this with @liggitt yesterday, this won't catch any strictJSONErrs if `into` is an `unstructured.Unstructured`, because Unstructured uses a custom unmarshaler that the decoder delegates to.

We need to unmarshall directly into the object.
See what I've done [here](https://github.com/kubernetes/kubernetes/pull/105916/files#diff-54b1d11ee7334f5492b87652fae8614bd77ce0fc0428f3ce986e9556fb1f9252R185-R187).

I like your approach of having a helper to reduce the duplication, but maybe we need to change the type of `into` so that it is `unmarshal(into interface{}, data, originalData []byte)`. Then up at line 165 we can do something like:

```
m := u.UnstructuredContent()
s.unarmshal(&m, data, originalData)
u.SetUnstructuredContent(m)
into = u
```

Will this work? Thoughts @liggitt?

We should probably have a test case for this situation too.

Reply:
Yes, I think your suggestion is good (`s.unmarshal` into `m`), but as far as I can tell that should only happen if `isUnstructured` is true, but not if `isUnstructured` is false and `runtime.IsNotRegisteredError(err)` is true, so the cases need to be split [here](https://github.com/kubernetes/kubernetes/pull/105822/files#diff-54b1d11ee7334f5492b87652fae8614bd77ce0fc0428f3ce986e9556fb1f9252R164). Test cases are indeed needed. I don't think `into = u` is needed as `u` is a cast and hence the same object as `into` already, and `SetUnstructuredContent` must be implemented on the pointer reciever.

Reply:
maybe add a comment here that this performs both duplicate and unknown field checking.

Reply:
agree this should be sufficient:
```
m := u.UnstructuredContent()
s.unmarshal(&m, data, originalData)
u.SetUnstructuredContent(m)
```

I don't think we need to split the IsNotRegisteredError and isUnstructured cases... we should do our own unstructured detection inside this function ... it should apply any time we're decoding strictly into an unstructured object, right?

agree we need testcases